### DISCUSSION TITLE
Allow restart of discovery/advertising without stopping existing connections

### DISCUSSION THREAD

**Discussion by jacoblubecki** [google/nearby#2874] [Ideas]
## Background

From the following documentation, it appears that advertising and discovery processes are intended to be one-time operations:
- [ConnectionsClient#startAdvertising](https://developers.google.com/android/reference/com/google/android/gms/nearby/connection/ConnectionsClient#public-abstract-taskvoid-startadvertising-byte[]-endpointinfo,-string-serviceid,-connectionlifecyclecallback-connectionlifecyclecallback,-advertisingoptions-options)
- [ConnectionsClient#startDiscovery](https://developers.google.com/android/reference/com/google/android/gms/nearby/connection/ConnectionsClient#public-abstract-taskvoid-startdiscovery-string-serviceid,-endpointdiscoverycallback-endpointdiscoverycallback,-discoveryoptions-options)

i.e. both methods mention:
> [ConnectionsStatusCodes.STATUS_OUT_OF_ORDER_API_CALL](https://developers.google.com/android/reference/com/google/android/gms/nearby/connection/ConnectionsStatusCodes#STATUS_OUT_OF_ORDER_API_CALL) if the app is currently connected to remote endpoints; call [stopAllEndpoints()](https://developers.google.com/android/reference/com/google/android/gms/nearby/connection/ConnectionsClient#public-abstract-void-stopallendpoints) first.

## Example

Using the collaborative whiteboard example from the nearby connections overview page, let's assume you wanted to use this in a university classroom setting. The only topology that could theoretically support this number of clients (say 20+) would be M-to-N, so from a connectivity standpoint, all nodes are equal peers.

If a student leaves to go to the water fountain and this breaks the mesh (e.g. because they were a single connecting node between two large subgraphs), you would want to be able to reconnect the subgraphs of students without either of the existing graphs collapsing, and also make it easy for the student to rejoin when they come back from the fountain.

Let's say the student that left was `B`, and the nodes they sat between were `A` and `C`.
1. `B` leaves, producing disconnected subgraphs represented by `A` (in the main graph) and `C` (in a satellite graph)
2. `A` and `C` have some localized info about the graph before `B` left, and `C` knows it can try to connect to `A` to reconnect the subgraphs
3. If `C` would like to connect to `A`, then `C` would need to call `stopAllEndpoints` which would disconnect them from the rest of the satellite graph containing `C`
4. This would propagate to the rest of the nodes in the satellite graph, and require all users to assimilate from scratch into the main graph

The main graph would also have to have at least one client advertising for assimilation to even be possible.
1. Lets say `A` wants to start advertising because it is now the only unsaturated node in the main graph (i.e. since `B` has left)
2. `A` has to call `stopAllEndpoints` to start advertising again
3.  This would cause the main graph containing `A` to collapse in the same way that the satellite graph containing `C` did

This seems to show that both subgraphs have to collapse and completely reorganize any time a key node drops out. It also seems to show that clients which become unsaturated cannot re-enable discovery without breaking the whole graph down, which sort of prohibits allowing other users to drop in after initial graph creation. Leaving discovery on forever is definitely not the answer based on the documentation heavily encouraging users to call `stopDiscovery` after some initial connections are made.

So it seems like having someone jump in to an existing connectivity session, or having someone leave and come back is practically very difficult if not impossible. Am I missing something? Is there a way to work around this issue?

I believe that clients _can_ attempt to connect to previously discovered nodes even if discovery is not active, but this seems very brittle since there's no way to actually be sure those nodes are even in range, let alone whether or not they are saturated and willing to accept a connection.

**Comment by jacoblubecki** [comment]
Following up to see if something like this is on the roadmap or what I might be able to contribute to help support. 

Also curious if there are any serious technical limitations to something like this.

**Comment by Xlythe** [comment]
There seems to be a few misconceptions.

Advertising, Discovery, and Connections are independent of each other. You can stop/start any of them without impacting the others. It's actually recommended that you don't continue to discover while connected if you can avoid it, as scanning will cause performance degradation for connections.

It looks like we never exposed ConnectionsClient#updateAdvertisingOptions and ConnectionsClient#updateDiscoveryOptions, which are methods we added to change what you're advertising/scanning without having to stop first. We can definitely fix that. But these are optimizations which patch small issues (such as race conditions if a remote device attempts to connect to you in the window that you stopped/restarted advertising, or an optimization so you don't need to rescan for discovered targets again if you stopped/restarted discovery), and in the meanwhile you can simply stop/restart to change AdvertisingOptions/DiscoveryOptions.

StopAllEndpoints is intended for when you are shutting down your application, and will stop Advertising, Discovery, and disconnect all ongoing connections. It's not intended for ongoing usage of Nearby Connections.

We don't support mesh networks, and the M to N topology is how many local incoming/outgoing connections are supported at once for a single device. Typically, devices will hit device resource limits after 3-5 connections, although if you're on the same WiFi network as the other devices this limitation is more on the router than the phone. You can build mesh networks on top of Nearby Connections, but you will need to solve for the cases you mentioned (where devices leave/rejoin).

**Reply by jacoblubecki** [reply]
So, to clarify -- it's totally possible to restart advertising and/or discovery _without_ calling `stopAllEndpoints`?

The documentation I am specifically concerned about is here (there is also a similar bullet point for `startDiscovery`):
<img width="960" alt="スクリーンショット 2024-11-09 午前11 23 38" src="https://github.com/user-attachments/assets/e1ae2ea5-961e-4409-a86a-f602de3841bf">

The way it's worded seems very explicit that the library will return a failure code if I try to start advertising again while connected.

> It's actually recommended that you don't continue to discover while connected if you can avoid it

This is why I am concerned -- if I can't keep discovery on forever and I also have to terminate all connections to restart it, then my implementation plan would have gone up in smoke.

I understand that nearby doesn't provide mesh functionality out of the box, but that it's theoretically possible with the M-to-N topology and that I would need to handle that logic on my own.

It just seemed impossible to make a practical mesh network if I had to call `stopAllEndpoints` any time the mesh needed to restructure itself somehow.

**Reply by Xlythe** [reply]
Correct, you don't need to call #stopAllEndpoints. Just calling #stopAdvertising is enough to allow you to later call #startAdvertising again.
